index.mjs

```diff
@@ -146,34 +146,39 @@ ParallelHook.prototype[$map] = function(f){
   return ParallelHook(map(f)(this.hook));
 }
 
-const crash = e => Future(() => { throw e });
+const CustomFuture = function (interpret) {
+  this._interpret = interpret;
+}
+
+CustomFuture.prototype = Object.create (Future.prototype);
+
+ParallelHook.prototype[$ap] = function(parallelFunctionHook){
+  const withValue = runHook (sequential (this));
+  const withFunction = runHook (sequential (parallelFunctionHook));
+  return ParallelHook(Hook(consumeResult => {
+    let result = null;
 
-ParallelHook.prototype[$ap] = function(mf){
-  return ParallelHook(Hook(c => {
-    let consume = noop;
-    const rf = mf.hook.run(f => consume !== noop ? consume (f) : (
-      Future((rej, res) => { consume = x => map(y => (res(y), y))(c(f(x))) })
+    const setResult = value => new CustomFuture (() => {
+      result = {value};
+      return noop;
+    });
```

Comment:
This Future causes its side-effect, and then never resolves.

Reply:
As I'm typing this explanation, I realize that I've introduced yet another bug. This time it will cause resources to never be disposed after they've been consumed. I'm going to have to fix that.